_posts/2020-11-03-torch-tabular/torch_tabular.Rmd

```diff
@@ -0,0 +1,445 @@
+    "stalk-color-above-ring",
+    "stalk-color-below-ring",
+    "veil-type",
+    "veil-color",
+    "ring-type",
+    "ring-number",
+    "spore-print-color",
+    "population",
+    "habitat"
+  ),
+  col_types = rep("c", 23) %>% paste(collapse = "")
+) %>%
+  # can as well remove because there's just 1 unique value
+  select(-`veil-type`)
+```
+
+In `torch`, `dataset()` creates an R6 class. Like with most R6 classes, there will usually be a need for an `initialize()` method. Below, we use `initialize()` to preprocess the data and store it in convenient pieces. More on that in a minute. Prior to that, please note the two other methods a `dataset` has to implement:
+
+-   `.getitem(i)` . This is the whole purpose of a `dataset`: Retrieve and return the observation located at some index it is asked for. Which index? That's to be decided by the caller, a `dataloader`. During training, usually we want to permute the order in which observations are used, while not caring about order in case of validation or test data.
+
+-   `.length()`. This method, again for use of a `dataloader`, indicates how many observations there are.
+
+In our example, both methods are straightforward to implement. `.getitem(i)` just directly uses its argument to index into the data, and `.length()` just returns the number of observations:
+
+```{r}
+mushroom_dataset <- dataset(
+  name = "mushroom_dataset",
+
+  initialize = function(indices) {
+    data <- self$prepare_mushroom_data(mushroom_data[indices, ])
+    self$xcat <- data[[1]][[1]]
+    self$xnum <- data[[1]][[2]]
+    self$y <- data[[2]]
+  },
+
+  .getitem = function(i) {
+    xcat <- self$xcat[i, ]
+    xnum <- self$xnum[i, ]
+    y <- self$y[i, ]
+    
+    list(x = list(xcat, xnum), y = y)
+  },
+  
+  .length = function() {
+    dim(self$y)[1]
+  },
+  
+  prepare_mushroom_data = function(input) {
+    
+    input <- input %>%
+      mutate(across(.fns = as.factor)) 
+    
+    target_col <- input$poisonous %>% 
+      as.integer() %>%
+      `-`(1) %>%
+      as.matrix()
+    
+    categorical_cols <- input %>% 
+      select(-poisonous) %>%
+      select(where(function(x) nlevels(x) != 2)) %>%
+      mutate(across(.fns = as.integer)) %>%
+      as.matrix()
+
+    numerical_cols <- input %>%
+      select(-poisonous) %>%
+      select(where(function(x) nlevels(x) == 2)) %>%
+      mutate(across(.fns = as.integer)) %>%
+      as.matrix()
+    
+    list(list(torch_tensor(categorical_cols), torch_tensor(numerical_cols)),
+         torch_tensor(target_col))
+  }
+)
+```
+
+As for data storage, there is a field for the target, `self$y`, but instead of the naturally to-be-expected `self$x` we see separate fields for numerical features (`self$xnum`) and categorical ones (`self$xcat`). This is just for convenience: The latter will be passed into embedding modules, which require its inputs to be of type `torch_long()`, as opposed to most other modules that, by default, work with `torch_float()`.
```

Comment:
```suggestion
As for data storage, there is a field for the target, `self$y`, but instead of the expected `self$x` we see separate fields for numerical features (`self$xnum`) and categorical ones (`self$xcat`). This is just for convenience: The latter will be passed into embedding modules, which require its inputs to be of type `torch_long()`, as opposed to most other modules that, by default, work with `torch_float()`.
```